pyccl/p2d.py

```diff
@@ -0,0 +1,144 @@
+        if(pkfunc is None):  # Initialize power spectrum from 2D array
+            # Make sure input makes sense
+            if (a_arr is None) or (lk_arr is None) or (pk_arr is None):
+                raise TypeError("If you do not provide a function, "
+                                "you must provide arrays")
+
+            pkflat = pk_arr.flatten()
+            # Check dimensions make sense
+            if (len(a_arr)*len(lk_arr) != len(pkflat)):
+                raise ValueError("Size of input arrays is inconsistent")
+        else:  # Initialize power spectrum from function
+            # Check that the input function has the right signature
+            try:
+                pkfunc(k=np.array([1E-2, 2E-2]), a=0.5)
+            except Exception:
+                raise TypeError("Can't use input function")
+
+            # Set k and a sampling from CCL parameters
+            nk = lib.get_pk_spline_nk()
+            na = lib.get_pk_spline_na()
+            a_arr, status = lib.get_pk_spline_a(na, status)
+            check(status)
+            lk_arr, status = lib.get_pk_spline_lk(nk, status)
+            check(status)
+
+            # Compute power spectrum on 2D grid
+            pkflat = np.zeros([na, nk])
+            for ia, a in enumerate(a_arr):
+                pkflat[ia, :] = pkfunc(k=np.exp(lk_arr), a=a)
+            pkflat = pkflat.flatten()
+
+        self.psp, status = lib.set_p2d_new_from_arrays(lk_arr, a_arr, pkflat,
+                                                       int(interp_order_lok),
+                                                       int(interp_order_hik),
+                                                       int(is_logp), status)
+        check(status)
+        self.has_psp = True
+
+    def eval(self, k, a, cosmo=None):
+        """Evaluate power spectrum.
+
+        Args:
+            k (float or array_like): wavenumber value(s) in units of Mpc^-1.
+            a (float): value of the scale factor
+            cosmo (:obj:`Cosmology`): Cosmology object. The cosmology object
+                  is needed in order to evaluate the power spectrum outside
+                  the interpolation range in `a`. E.g. if you want to evaluate
+                  the power spectrum at a very small a, not covered by the
+                  arrays you passed when initializing this object, the power
+                  spectrum will be extrapolated from the earliest available
+                  value using the linear growth factor (for which a cosmology
+                  is needed).
+            a_arr (array): an array holding values of the scale factor.
+
+        Returns:
+            float or array_like: value(s) of the power spectrum.
+        """
+        status = 0
+        if cosmo is not None:
+            cospass = cosmo.cosmo
+        else:
+            raise NotImplementedError("Currently we need a cosmology to "
+                                      "extrapolate growth")
+            cospass = None
```

Comment:
This block means you have to pass a cosmology even though the parameter argument is optional.

Reply:
Made non-optional